### PR TITLE
overlord,snap: add support for per-snap storage on ubuntu-save

### PR DESCRIPTION
**Pull request by Meulengracht**
Support for storing per-snap data onto ubuntu-save. This PR contains the snapd side and relies on early mount of the ubuntu-save partition to be mounted under /var/lib/snapd/save. The support for this is here for core22 (https://github.com/snapcore/core-base/pull/49), and the one for core20 (https://github.com/snapcore/core20/pull/134)

This PR also needs https://github.com/snapcore/snapd/pull/11933 to pass it's tests

### CODE REVIEW

**Review by pedronis — APPROVED** [review]
thank you

File: overlord/snapstate/backend/copydata.go [hunk 1]
```diff
@@ -108,6 +109,40 @@ func (b Backend) UndoCopySnapData(newInfo, oldInfo *snap.Info, _ progress.Meter,
...
+	// that it is indeed a mount-point.
+	if hasSave, err := osutil.IsMounted(dirs.SnapSaveDir); err != nil || !hasSave {
+		if err != nil {
+			return fmt.Errorf("cannot check if ubuntu-save is mounted: %v", err)
+		}
+		return nil
+	}
+
+	saveDir := snap.CommonDataSaveDir(info.InstanceName())
+	return os.MkdirAll(saveDir, 0755)
```
Comment (mardy) [inline overlord/snapstate/backend/copydata.go:128]: I think that this makes sense, but I'm writing just to make sure it's intended (maybe a comment could be added to make this explicit): if the permission is 0755, it means that only the root use can create files under ubuntu-save.

File: overlord/snapstate/backend/copydata_test.go [hunk 2]
```diff
@@ -582,6 +583,108 @@ func (s *copydataSuite) TestUndoCopyDataSameRevision(c *C) {
...
+	defer restore()
+
+	v1 := snaptest.MockSnap(c, helloYaml1, &snap.SideInfo{Revision: snap.R(10)})
+
+	// first install
+	err := s.be.SetupSnapSaveData(v1, progress.Null)
+	c.Assert(err, IsNil)
+	_, err = os.Stat(v1.CommonDataSaveDir())
+	c.Assert(err, IsNil)
+
```
Comment (mardy) [inline overlord/snapstate/backend/copydata_test.go:629]: At this point, please create a file under the save dir, to make sure that it gets removed later too.

**Review by mardy — APPROVED** [review]
LGTM, thanks! Just a few nitpicks and questions.

File: dirs/dirs.go [hunk 3]
```diff
@@ -408,6 +409,7 @@ func SetRootDir(rootdir string) {
 	SnapFDEDir = SnapFDEDirUnder(rootdir)
 	SnapSaveDir = SnapSaveDirUnder(rootdir)
 	SnapDeviceSaveDir = filepath.Join(SnapSaveDir, "device")
+	SnapDataSaveDir = filepath.Join(SnapSaveDir, "snap")
```
[the remark this answers is not in the document]
Comment (pedronis) [reply dirs/dirs.go:412]: @mardy "snap" is correct, is the same as /var/snap. It was a typo on my part, I also pointed to the spec which has the right name. If we communicated snaps to a costumers or put it into snap-declarations we need to see to correct that.

[the remark this answers is not in the document]
Comment (mardy) [reply dirs/dirs.go:412]: Yes we did :-) I'll write them about using "snap" instead.

File: tests/nested/core/save-data/task.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    echo "Ensuring the path /var/lib/snapd/save/snap/test-snapd-sh now exists"
+    tests.nested exec "ls -l /var/lib/snapd/save/snap/test-snapd-sh"
+
+    # instance environment variables are correctly set up
+    tests.nested exec "snap run test-snapd-sh.sh -c 'env' test > snap_foo-env.txt"
+    tests.nested exec "grep 'SNAP_SAVE_DATA=/var/lib/snapd/save/snap/test-snapd-sh' < snap_foo-env.txt"
+
+    echo "Ensuring we can write a file to /var/lib/snapd/save/snap/test-snapd-sh"
+    tests.nested exec "sudo snap run test-snapd-sh.sh -c \"echo 'hello world' > /var/lib/snapd/save/snap/test-snapd-sh/hello.txt\""
+
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply tests/nested/core/save-data/task.yaml:36]: Good point, let me extend the test!

File: overlord/snapstate/backend/copydata_test.go [hunk 5]
```diff
@@ -582,6 +583,108 @@ func (s *copydataSuite) TestUndoCopyDataSameRevision(c *C) {
...
+	// removes correctly when no previous info is present
+	err = s.be.UndoSetupSnapSaveData(v1, nil, progress.Null)
+	c.Assert(err, IsNil)
+	_, err = os.Stat(v1.CommonDataSaveDir())
+	c.Check(os.IsNotExist(err), Equals, true)
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply overlord/snapstate/backend/copydata_test.go:634]: Good idea, thanks!

File: overlord/snapstate/backend/copydata.go (hunk 1 above)
Comment (Meulengracht) [reply overlord/snapstate/backend/copydata.go:128]: I probably just need @pedronis to verify this should be like this, my understanding is that this is the correct behaviour for writing files to the folder on ubuntu-save, but I would love a second thumbs up on this

File: overlord/snapstate/backend/copydata_test.go (hunk 2 above)
Comment (Meulengracht) [reply overlord/snapstate/backend/copydata_test.go:629]: Thank you, you are right! Fixed

File: dirs/dirs.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Meulengracht) [reply dirs/dirs.go:412]: I'm sorry @mardy if I communicated this wrong to you at some point, I wouldn't be surprised if I either wrote it wrong or didn't notice the extra 's'

File: overlord/snapstate/backend/copydata.go (hunk 1 above)
Comment (pedronis) [reply overlord/snapstate/backend/copydata.go:128]: should behave like /var/snap/common which I think is also like this

**Pull request merged**